File: extended-gcd/once.c
#include <stdio.h>
#include <math.h>

long long int min(long long int a,long long int b)
{
    if(a<b)
    {
        return a;
    }
    else
    {
        return b;
    }
}


long long int max(long long int a,long long int b)
{
    if(a>b)
    {
        return a;
    }
    else
    {
        return b;
    }
}


long long int gcd(long long int a,long long int b)
{
    if(b==0)
    {
        return a;
    }
    return gcd(b,a%b);
}

long long int inv(long long int a,long long int b,long long int *x,long long int *y)
{
    if(b==0)
    {
        *x=1;
        *y=0;
        return a;
    }
    long long int x1,y1;
    long long int d= inv(b,a%b,&x1,&y1);
    *x=y1;
    *y=x1-y1*(a/b);
    return d;
}


int main()
{
    while(1)
    {
        long long int a1,a2,b1,b2;
        scanf("%lld %lld %lld %lld",&a1,&b1,&a2,&b2);
        if(a1+a2+b1+b2==0)
        {
            break;
        }
        long long int c=a2-a1;
        long long int x,y;
        long long int g=inv(b1,b2,&x,&y);
        if(c%g!=0)
        {
            printf("Impossible\n");
            continue;
        }
        y*=-1;
        //x=(c/g)*(x+k*b2/g)>0;
        //y=(c/g)*(y+k*b1/g)>=0;
        long long int big=1000000000000000000;
        long long int mink=-big;
        long long int maxk=big;
        //x = (c/g)*(x+k*b2/g)>0;
        printf("x=%lld y=%lld\n",x,y);
        if(c>0)
        {
            mink = -x/b2 + 1;
        }
        else if(c<0)
        {
            if(x%b2)
            {
                maxk= -x/b2;
            }
            else
            {
                maxk = -x/b2 -1;
            }
        }
        //y = (c/g)*(y+k*b1/g)>=0;
        if(c>0)
        {
            if(y%b1)
            {
                mink = max(mink,-y/b1 + 1);
            }
            else
            {
                mink = max(mink,-y/b1);
            }
        }
        else
        {
            maxk = min(maxk,-y/b1);
        }
        printf("mink=%lld maxk=%lld\n",mink,maxk);
        long long int k;
        if(mink==-big)
        {
            k=maxk;
        }
        else if (maxk==big)
        {
            k=mink;
        }
        else if (mink<=maxk)
        {
            int cc;
        }
        else
        {
            printf("Impossible\n");
            continue;
        }
        printf("%lld %lld\n",a1+b1*((c/g)*(x + k*b2/g)),a2+b2*((c/g)*(y + k*b1/g)));
    }
}

File: extended-gcd/once.py
import math

def inv(a, b): 
	if b == 0 : 
		return a,1,0
			
	gcd,x1,y1 = inv(b, a%b) 
	y = x1 - (a//b) * y1 
	x = y1 	
	return gcd,x,y 

while(1):
    inp = input().split(' ')
    a1=int(inp[0])
    b1=int(inp[1])
    a2=int(inp[2])
    b2=int(inp[3])
    if a1+a2+b2+b1 == 0:
        break
    c=a2-a1
    if(b1==0 and b2):
        if((a1-a2) %b2 ):
            print("Impossible")
        else:
            if((a1-a2)/b2 >=0):
                print(a1)
            else:
                print("Impossible")
        continue
    elif(b2==0 and b1):
        if((a2-a1) %b1 ):
            print("Impossible");
        else:
            if((a2-a1)/b1 >0):
                print(a2)
            else:
                print("Impossible")
        continue
    elif(b2+b1 == 0):
        if a1 == a2:
            print(a1)
        else:
            print("Impossible")
        continue
    g,x,y=inv(b1,b2)
    if(c%g):
        print("Impossible")
        continue
    y*=-1
    # print(g,x,y)
    big=1000000000000000000
    mink=-big
    maxk=big
    x*=(c/g)
    y*=(c/g)

    '''
    mink = math.floor(-x*g/b2) + 1
    if y*g%b1:
        mink = max(mink,math.floor(-y*g/b1)+1)
    else:
        mink = max(mink,math.floor(-y*g/b1))
    '''
    mink = math.floor(-y*g/b1) + 1
    if x*g%b2:
        mink = max(mink,math.floor(-x*g/b2)+1)
    else:
        mink = max(mink,math.floor(-x*g/b2))
       
    # print(x,y)
    # print(mink,x+mink*b2/g,y+mink*b1/g)
    print(int(a1+b1*(x + mink*b2/g)))



    '''
    # x = (c/g)*(x+k*b2/g)>0;
    if c>0:
        mink = int(-x/b2) + 1
        if y%b1:
            mink = max(mink,int(-y/b1)+1)
        else:
            mink = max(mink,int(-y/b1))
        #print(mink,((c/g)*(x + mink*b2/g)),((c/g)*(y + mink*b1/g)))
       # print(a1+b1*((c/g)*(x + mink*b2/g)),a2+b2*((c/g)*(y + mink*b1/g)))
        print(int(a1+b1*((c/g)*(x + mink*b2/g))))
    elif c<0:
        if(x%b2):
            maxk = math.floor(-x/b2)
        else:
            maxk = int(-x/b2) -1
        maxk = min(maxk,math.ceil(-y/b1))
       # print(maxk,((c/g)*(x + maxk*b2/g)),((c/g)*(y + maxk*b1/g)))
       # print(a1+b1*((c/g)*(x + maxk*b2/g)),a2+b2*((c/g)*(y + maxk*b1/g)))
        print(int(a1+b1*((c/g)*(x + maxk*b2/g))))
    '''